File: bot.py
import discord
import os
import re
import datetime
import random
import json
import asyncio
from discord.ui import Button, View 
from dotenv import load_dotenv

print(discord.__version__)

load_dotenv()  # Load environment variables from a .env file (explained below)
TOKEN = os.getenv('DISCORD_TOKEN')  # Get your bot token
game_in_progress = False

# Specify Intents
intents = discord.Intents.default()  # For basic functionality
intents.message_content = True  # Make sure to include message_content 

client = discord.Client(intents=intents)

@client.event
async def on_ready():
  print(f'{client.user} has connected to Discord!')

@client.event
async def on_message(message):
    if message.author == client.user:  # Prevent the bot from responding to itself
        return
    if game_in_progress:  # You'll need to define this function 
        if message.content.startswith('!resign') or message.content.startswith('!ff'):
            await handle_resignation(message)  # Define handle_resignation function
            await message.delete()
        return  # Ignore all other commands during a game 
    elif message.content.startswith('!thursday'):
        await create_weekly_calendar_message(message.channel, day_offset=3)  # Send to trigger channel
        await message.delete()  # Delete the trigger message
    elif message.content.startswith('!friday'): 
        await create_weekly_calendar_message(message.channel, day_offset=4)  # Send to trigger channel
        await message.delete()  # Delete the trigger message
    elif message.content.startswith('!triviastats'): 
        await calculate_trivia_stats(message.channel) 
        await message.delete() 
    elif message.content.startswith('!herodraft3'):
        await start_herodraft(message, draft_size=3) 
    elif message.content.startswith('!herodraft5'):
        await start_herodraft(message, draft_size=5)  
    elif message.content.startswith('!roll'):
        await calculate_roll_result(message.content[5:], message)

@client.event
async def start_herodraft(message, draft_size):
    button = Button(label="Accept Challenge", style=discord.ButtonStyle.green)
    view = View(timeout=60)  # Timeout the challenge after 60 seconds
    view.add_item(button) 

    challenge_msg = await message.channel.send(f"{message.author.mention} has issued a Hero Draft challenge!", view=view)

    async def button_callback(interaction):  # Define callback for button press
        global game_in_progress  # Indicate game is now starting
        game_in_progress = True

        challenger = message.author
        opponent = interaction.user

        if challenger == opponent:  # Check for self-battle attempt
            await interaction.response.send_message("You cannot battle yourself!", ephemeral=True)
            return  # Prevent the game from starting

        # Embed is better for game start announcements
        embed = discord.Embed(title="Game Starting!", 
                              description=f"{challenger.mention} vs {opponent.mention}")
        await interaction.response.edit_message(content=None, embed=embed, view=None)  # Remove the button 

        # Start the game logic (replace with your game-specific function)
        await herodraft(challenger, opponent, message.channel)

    button.callback = button_callback  # Assign the callback to the button

async def herodraft(challenger, opponent, channel):
    await channel.send(f"The game between {challenger.mention} and {opponent.mention} begins!")
    current_player = challenger
    challenger_health = 3
    opponent_health = 3

    if random.randint(0, 1) == 0:
        current_player = challenger
        await channel.send(f"{challenger.mention} won the coin toss and goes first!")
    else:
        current_player = opponent
        await channel.send(f"{opponent.mention} won the coin toss and goes first!")

    while True:  # Main game loop
        embed = discord.Embed(title=f"{current_player.mention}'s Turn",
                              description=f"**Challenger:** {challenger_health} HP\n"
                                          f"**Opponent:** {opponent_health} HP")

        attack_button = discord.ui.Button(label="Attack", style=discord.ButtonStyle.red)
        pass_button = discord.ui.Button(label="Pass", style=discord.ButtonStyle.grey)
        view = discord.ui.View()  # Create a View to hold the buttons
        view.add_item(attack_button)
        view.add_item(pass_button)

        await channel.send(embed=embed, view=view)

        async def button_callback(interaction):
            nonlocal challenger_health, opponent_health, current_player 

            if interaction.user != current_player:
                await interaction.response.send_message("It's not your turn!", ephemeral=True)
                return

            if interaction.component.label == "Attack":
                if current_player == challenger:
                    opponent_health -= 1
                else:
                    challenger_health -= 1

                await interaction.response.edit_message(content=f"{interaction.user.mention} attacked!", view=None)
            
            if interaction.component.label == "Pass":
                await interaction.response.edit_message(content=f"{interaction.user.mention} passed their turn.", view=None)

            # Switch turns
            current_player = opponent if current_player == challenger else challenger 

            # Check for win condition
        if opponent_health <= 0:
            await channel.send(f"{challenger.mention} wins!")
            break
        elif challenger_health <= 0:
            await channel.send(f"{opponent.mention} wins!")
            break

        attack_button.callback = button_callback
        pass_button.callback = button_callback

        challenge_msg = await message.channel.send(f"{message.author.mention} has issued a Hero Draft challenge!", view=view)

        try:
            interaction = await client.wait_for('interaction', check=lambda i: i.message.id == challenge_msg.id, timeout=60)  
            await button_callback(interaction)  # Process the interaction
        except asyncio.TimeoutError:
            await channel.send("Challenge timed out due to inactivity.") 

@client.event
async def handle_resignation(message):
    global game_in_progress
    game_in_progress = False
@client.event
async def calculate_roll_result(roll_string, message):
    dice_pattern = re.compile(r'(\d+d\d+)(?:\s*(\+|-)\s*(\d+d\d+|\d+))?')  # Modified pattern
    longMess = False
    rolls = []
    total = 0
    difficulty = 1
    maxroll = 0

    for part in dice_pattern.findall(roll_string):
        num_dice, die_type = map(int, part[0].split('d'))
        difficulty = difficulty * num_dice * die_type  # Calculate difficulty
        maxroll += num_dice * die_type
        if(num_dice > 1):
            longMess = True
        rolls.append([random.randint(1, die_type) for _ in range(num_dice)])
        total += sum(rolls[-1])

        if part[1]:  # If a modifier exists 
            if part[1] == '+':
                if part[2].isdigit():  # Check for a flat number
                    maxroll += int(part[2])
                    total += int(part[2])
                else: 
                    num_dice, die_type = map(int, part[2].split('d'))
                    difficulty = difficulty * num_dice * die_type
                    maxroll += num_dice * die_type
                    rolls.append([random.randint(1, die_type) for _ in range(num_dice)])
                    total += sum(rolls[-1])
            elif part[1] == '-':
                if part[2].isdigit(): 
                    maxroll -= int(part[2])
                    total -= int(part[2])
                else: 
                    num_dice, die_type = map(int, part[2].split('d'))
                    difficulty = difficulty * num_dice * die_type
                    maxroll += num_dice * die_type
                    rolls.append([random.randint(1, die_type) for _ in range(num_dice)])
                    total -= sum(rolls[-1])

    output = f"{message.author.mention} rolled "
    if len(rolls) > 1:
        longMess = True

    if longMess:  
        output += f"({', '.join(map(str, sum(rolls, [])))}) "  
        output += f"Total: {total}!" 
    else:
        output += f"{total}!"

    if difficulty >= 20:
        if total == maxroll:
            output += " CRITICAL SUCCESS!"
        elif total == 1 * len(rolls):  # Check if each die rolled its minimum value
            output += " Critical Failure..."
          
    await message.channel.send(output) 
@client.event
async def calculate_trivia_stats(channel):
    user_scores = {}

    async for message in channel.history(limit=10):  # Fetch all history
        if message.author.equals('trivia-bot'):  # Check if it's the trivia bot's response
            print('bot message:', message.content)
            referenced_message = await message.channel.fetch_message(message.reference.message_id)
            print('ref message:', referenced_message)
            username = referenced_message.content.split()[0]  # Extract username from '/question'

            if username not in user_scores:
                user_scores[username] = {'correct': 0, 'total': 0} 

            user_scores[username]['total'] += 1
            if 'Correct!' in message.content:
                user_scores[username]['correct'] += 1

    # Build the stats message
    stats_message = "Trivia Stats:\n"
    for username, scores in user_scores.items():
        accuracy = round(scores['correct'] / scores['total'] * 100, 1) if scores['total'] else 0
        stats_message += f"- {username}: {scores['correct']}/{scores['total']} ({accuracy}%)\n"  # Indent

    print(stats_message)
@client.event
async def create_weekly_calendar_message(channel, start_timestamp=None, day_offset=3):
    """
    Creates and sends a weekly calendar message to the specified channel.

    Args:
        channel (discord.TextChannel): The channel to send the message to.
        session_number (int, optional): The starting session number. Defaults to 13.
        start_timestamp (int, optional): A Unix timestamp to start from. If not
                                        provided, the next Thursday will be used.
    """
    async def find_session_number(channel):
        """Fetches the most recent session number from channel history"""
        async for msg in channel.history(limit=50):  
            if msg.content.startswith("Session "):
                parts = msg.content.split()  
                if len(parts) >= 2:
                    session_str = parts[1]
                    if session_str.endswith(":"):  # Check for the colon
                        session_str = session_str[:-1]  # Remove the colon
                    if session_str.isdigit():
                        return int(session_str)
                return None  
        return None  # No session message found

    session_number = await find_session_number(channel)
    if session_number is None:
        session_number = 13  # Default starting session number

    # Calculate timestamps for next Thursday
    if start_timestamp is None:
        now = datetime.datetime.now()
        day_adjustment = (day_offset - now.weekday() + 7) % 7  # Adjusted calculation
        target_day = now + datetime.timedelta(days=day_adjustment)
        start_timestamp = int(target_day.replace(hour=18, minute=0, second=0).timestamp())
        end_timestamp = int(target_day.replace(hour=21, minute=0, second=0).timestamp())


    # Create the message content
    message = f"""
    Session {session_number+1}: 
    Next Session: <t:{start_timestamp}> - <t:{end_timestamp}:t>
    which is <t:{start_timestamp}:R>!
    """

    # Send the message
    await channel.send(message)

if TOKEN is not None:
  client.run(TOKEN)
else:
  print("No token provided.")

File: bot.js
// Install necessary libraries (using npm or yarn)
const Discord = require('discord.js');
const dotenv = require('dotenv');

// ... Environment Variable Setup (dotenv) ...

const client = new Discord.Client({ intents: [Discord.GatewayIntentBits.Guilds, Discord.GatewayIntentBits.GuildMessages, Discord.GatewayIntentBits.MessageContent] });
let gameInProgress = false;

client.on('ready', () => {
  console.log(`Logged in as ${client.user.tag}!`);
});

client.on('messageCreate', async (message) => { // Updated event name
  if (message.author.bot) return; 

  if (gameInProgress) {
    if (message.content.startsWith('!resign') || message.content.startsWith('!ff')) {
      await handleResignation(message);
      await message.delete();
    }
    return; 
  } else if (message.content.startsWith('!thursday')) {
    await createWeeklyCalendarMessage(message.channel, 3); 
    await message.delete(); 
  } else if (message.content.startsWith('!friday')) {
    await createWeeklyCalendarMessage(message.channel, 4); 
    await message.delete();
  } else if (message.content.startsWith('!triviastats')) {
    await calculateTriviaStats(message.channel);
    await message.delete();
  } else if (message.content.startsWith('!herodraft3')) {
    await startHerodraft(message, 3);
  } else if (message.content.startsWith('!herodraft5')) {
    await startHerodraft(message, 5);
  } else if (message.content.startsWith('!roll')) {
    await calculateRollResult(message.content.slice(5), message); // slice to remove '!roll'
  } 
});

async function startHerodraft(message, draftSize) {
  const button = new Discord.ButtonBuilder()
      .setLabel('Accept Challenge')
      .setStyle(Discord.ButtonStyle.Success)
      .setCustomId('accept'); // You'll need an interaction handler

  const view = new Discord.ActionRowBuilder()
      .addComponents(button); 

  const challengeMsg = await message.channel.send({ 
      content: `${message.author.toString()} has issued a Hero Draft challenge!`, 
      components: [view] 
  });

  // ... Interaction Handling (explained later) ...
}

async function herodraft(challenger, opponent, channel) {
  await channel.send(`The game between ${challenger.toString()} and ${opponent.toString()} begins!`);

  let currentPlayer = challenger;
  let challengerHealth = 3;
  let opponentHealth = 3;

  if (Math.random() < 0.5) {
    currentPlayer = challenger;
    await channel.send(`${challenger.toString()} won the coin toss and goes first!`);
  } else {
    currentPlayer = opponent;
    await channel.send(`${opponent.toString()} won the coin toss and goes first!`);
  }

  while (true) { 
    const embed = new Discord.EmbedBuilder()
      .setTitle(`${currentPlayer.toString()}'s Turn`)
      .setDescription(`**Challenger:** ${challengerHealth} HP\n**Opponent:** ${opponentHealth} HP`);

    // ... Create Attack/Pass buttons with custom IDs ...

    const view = new Discord.ActionRowBuilder()
      .addComponents(/* your buttons */);

    const sentMessage = await channel.send({ embeds: [embed], components: [view] });

    // ... Interaction Handling (explained below) ...
  }
}

async function calculateRollResult(rollString, message) {
    // Regular Expression (mostly unchanged)
    const dicePattern = /(\d+d\d+)(?:\s*(\+|-)\s*(\d+d\d+|\d+))?/g; 
  
    let longMess = false;
    const rolls = [];
    let total = 0;
    let difficulty = 1;
    let maxroll = 0;
  
    // Find all dice patterns using the regular expression
    let match;
    while ((match = dicePattern.exec(rollString)) !== null) {
      const [_, numDiceStr, operator, modifierStr] = match; // Destructure the match results
      const numDice = parseInt(numDiceStr.split('d')[0]);
      const dieType = parseInt(numDiceStr.split('d')[1]);
  
      difficulty *= numDice * dieType;
      maxroll += numDice * dieType;
  
      if (numDice > 1) {
        longMess = true;
      }
  
      // Generate individual die rolls
      const roll = []; 
      for (let i = 0; i < numDice; i++) {
        roll.push(Math.floor(Math.random() * dieType) + 1); 
      }
      rolls.push(roll);
      total += roll.reduce((sum, val) => sum + val, 0); // Calculate the sum of the roll
  
      // Process Modifier (if present)
      if (operator) {
        if (operator === '+' && modifierStr.match(/^\d+$/)) { // Check for flat number modifier
          maxroll += parseInt(modifierStr);
          total += parseInt(modifierStr);
        } else if (modifierStr.match(/\d+d\d+/)) {      // Check for another dice roll modifier
          const [modNumDiceStr, modDieTypeStr] = modifierStr.split('d');
          const modNumDice = parseInt(modNumDiceStr);
          const modDieType = parseInt(modDieTypeStr);
  
          difficulty *= modNumDice * modDieType;
          maxroll += modNumDice * modDieType;
  
          const modRoll = [];
          for (let i = 0; i < modNumDice; i++) {
            modRoll.push(Math.floor(Math.random() * modDieType) + 1);
          }
          rolls.push(modRoll);
  
          if (operator === '+') {
            total += modRoll.reduce((sum, val) => sum + val, 0);
          } else {
            total -= modRoll.reduce((sum, val) => sum + val, 0);
          }
        }
      }
    }
  
    // Construct Output Message
    let output = `${message.author.toString()} rolled `; 
  
    if (longMess) {
      output += `(${rolls.flat().join(', ')}) `; // Flatten the 'rolls' array
    }
    output += `Total: ${total}!`;
  
    // Critical Success/Failure
    if (difficulty >= 20) {
      if (total === maxroll) {
        output += " CRITICAL SUCCESS!";
      } else if (total === rolls.length) { // Note the slight change here
        output += " Critical Failure...";
      }
    }
  
    await message.channel.send(output);
}

async function calculateTriviaStats(channel) {
    const userScores = {};
    const triviaBotId = '...'; // Replace '...' with the actual ID of your Trivia Bot

    try {
        const messages = await channel.messages.fetch({ limit: 10 }); // Fetch recent messages

        for (const message of messages.values()) {
        if (message.author.id !== triviaBotId) continue; // Check if it's the trivia bot

        const referencedMessage = await message.fetchReference(); 
        if (!referencedMessage) continue; 

        const username = referencedMessage.content.split(' ')[0];

        if (!userScores[username]) {
            userScores[username] = { correct: 0, total: 0 };
        }

        userScores[username]['total']++;
        if (message.content.includes('Correct!')) {
            userScores[username]['correct']++;
        }
        }

        // Build the stats message
        let statsMessage = "Trivia Stats:\n";
        for (const [username, scores] of Object.entries(userScores)) { // Using Object.entries()
        const accuracy = scores['total'] ? round(scores['correct'] / scores['total'] * 100, 1) : 0;
        statsMessage += `- ${username}: ${scores['correct']}/${scores['total']} (${accuracy}%)\n`;
        }

        console.log(statsMessage);
        // You'll likely want to send the `statsMessage` to the channel: 
        // await channel.send(statsMessage); 

    } catch (error) {
        console.error("Error fetching or processing messages:", error);
    }
}

async function createWeeklyCalendarMessage(channel, startTimestamp = null, dayOffset = 3) {
    /**
     * Fetches the most recent session number from channel history
     */
    async function findSessionNumber(channel) {
        try {
        const messages = await channel.messages.fetch({ limit: 50 });
        for (const message of messages.values()) {
            if (message.content.startsWith("Session ")) {
            const parts = message.content.split(" ");
            if (parts.length >= 2) {
                let sessionStr = parts[1];
                if (sessionStr.endsWith(":")) {
                sessionStr = sessionStr.slice(0, -1); // Remove the colon
                }
                if (/\d+/.test(sessionStr)) { // Simple check if it's a number
                return parseInt(sessionStr);
                }
            }
            }
        }
        return null; // No session message found
        } catch (error) {
        console.error("Error fetching session number:", error);
        return null;
        }
    }
    
    let sessionNumber = await findSessionNumber(channel);
    if (sessionNumber === null) {
        sessionNumber = 13; // Default starting session number
    }
    
    // Calculate timestamps for next Thursday
    if (startTimestamp === null) {
        const now = new Date();
        const dayAdjustment = (dayOffset - now.getDay() + 7) % 7; 
        const targetDay = new Date(now.getTime() + dayAdjustment * 24 * 60 * 60 * 1000);
        targetDay.setHours(18, 0, 0); // Set to 6 PM
    
        startTimestamp = Math.floor(targetDay.getTime() / 1000); // Unix timestamp in seconds
        endTimestamp = startTimestamp + 3 * 60 * 60; // Add 3 hours
    }
    
    // Create the message content (Discord timestamp formatting might need an alternative)
    const message = `
    Session ${sessionNumber + 1}: 
    Next Session: <t:${startTimestamp}> - <t:${endTimestamp}:t>
    which is <t:${startTimestamp}:R>! 
    `;
    
    // Send the message
    await channel.send(message);
}

function round(number, decimalPlaces) {
const factor = 10 ** decimalPlaces;
return Math.round(number * factor) / factor;
}